lib/ansible/modules/cloud/amazon/cloudfront_facts.py

```diff
@@ -0,0 +1,448 @@
+        distribution_id = service_mgr.get_distribution_id_from_domain_name(domain_name_alias)
+        if not distribution_id:
+            module.fail_json(msg='Error unable to source a distribution id from domain_name_alias')
+
+    # set appropriate ansible_facts id
+    if distribution_id:
+        result = { 'ansible_facts': { 'cloudfront': { distribution_id:{} } } }
+        facts = result['ansible_facts']['cloudfront'][distribution_id]
+    elif cloud_front_origin_access_identity_id:
+        result = { 'ansible_facts': { 'cloudfront': { cloud_front_origin_access_identity_id:{} } } }
+        facts = result['ansible_facts']['cloudfront'][cloud_front_origin_access_identity_id]
+    elif web_acl_id:
+        result = { 'ansible_facts': { 'cloudfront': { web_acl_id:{} } } }
+        facts = result['ansible_facts']['cloudfront'][web_acl_id]
+    else:
+        result = { 'ansible_facts': { 'cloudfront': {} } }
+        facts = result['ansible_facts']['cloudfront']
+
+    # call details based on options
+    if distribution:
+        facts['distribution'] = service_mgr.get_distribution(distribution_id)
+    if distribution_config:
+        facts['distribution_config'] = service_mgr.get_distribution_config(distribution_id)
+    if cloud_front_origin_access_identity:
+        facts['cloud_front_origin_access_identity'] = service_mgr.get_cloud_front_origin_access_identity(cloud_front_origin_access_identity_id)
+    if cloud_front_origin_access_identity_config:
+        facts['cloud_front_origin_access_identity_config'] = service_mgr.get_cloud_front_origin_access_identity_config(cloud_front_origin_access_identity_id)
+    if invalidation:
+        facts['invalidation'] = service_mgr.get_invalidation(distribution_id, invalidation_id)
+    if streaming_distribution:
+        facts['streaming_distribution'] = service_mgr.get_streaming_distribution(distribution_id)
+    if streaming_distribution_config:
+        facts['streaming_distribution_config'] = service_mgr.get_streaming_distribution_config(distribution_id)
+
+    # call list based on options
+    if all_lists or list_cloud_front_origin_access_identities:
+        facts['list_cloud_front_origin_access_identities'] = service_mgr.list_cloud_front_origin_access_identities()
+    if all_lists or list_distributions:
+        facts['list_distributions'] = service_mgr.list_distributions()
+    if all_lists or list_streaming_distributions:
+        facts['list_streaming_distributions'] = service_mgr.list_streaming_distributions()
+    if list_distributions_by_web_acl_id:
+        facts['list_distributions_by_web_acl'] = service_mgr.list_distributions_by_web_acl()
+    if list_invalidations:
+        facts['list_invalidations'] = service_mgr.list_invalidations(distribution_id)
+
+    result['changed'] = False
+    module.exit_json(msg="Retrieved cloudfront facts.", ansible_facts=result)
+
+# import module snippets
+from ansible.module_utils.basic import *
+from ansible.module_utils.ec2 import *
```

Comment:
I think that wildcard imports are generally discouraged, because it makes it difficult to refactor and determine the origin of a function or class. For example, at first the `boto3_conn` function used in the constructor seemed undefined.